crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -815,29 +809,49 @@ where
     Ok(async move {
         let archiver = match maybe_archiver {
             Some(archiver) => archiver,
-            None => initialize_archiver(&segment_headers_store, &subspace_link, client.as_ref())?,
+            None => {
+                let Some(last_segment_header) = segment_headers_store.last_segment_header() else {
+                    unreachable!("Already checked above that max_segment_index is not None")
+                };
+                let last_archived_block_number =
+                    last_segment_header.last_archived_block().number.into();
+                let mut block_number_to_archive = Zero::zero();
+                // wait for unknown blocks to arrive.
+                while let Some(ref block_import_notification) =
+                    block_importing_notification_stream.next().await
+                {
+                    let block_number = block_import_notification.block_number;
+                    if block_number < last_archived_block_number {
+                        debug!(%block_number, %last_archived_block_number, "skip already know block");
+                        continue;
+                    }
+                    if block_number == last_archived_block_number {
+                        debug!(%block_number, "hit an edge of the last archived block");
+                        let best_block_hash = client.info().best_hash;
+                        block_number_to_archive = last_archived_block_number
+                            + client
+                                .runtime_api()
+                                .chain_constants(best_block_hash)?
+                                .confirmation_depth_k()
+                                .into();
```

Comment:
Looks like you can compute this before entering the loop and simplify code to a single `if` condition instead of three?

Also there is `subspace_link.chain_constants`, you don't need to query it again.